Lookup Fixture Missing on Restore Bug Fix

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
There exists a bug where if a fixture is configured with an index, the first related `schema` xml included in the restore but all following elements (including the fixture itself) is not. The resulted in app not being able to find the fixture. This PR fixes this bug.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Context for this PR:
There was a bug introduced [USH-5912](https://dimagi.atlassian.net/browse/USH-5912) as part of a [PR1](https://github.com/dimagi/commcare-hq/pull/36091). PR1 was partially reverted in [PR2](https://github.com/dimagi/commcare-hq/pull/36227). The functionalities reverted were: 1. filtering out for only last modified look up tables when constructing xml for the restore 2. caching fixtures individually instead of caching all global fixtures together.

Cause of the bug:
When getting the fixtures related to the data type, it was incorrectly returning only the first element in the list of elements. As a result, only the first element was written to the io stream.

This PR reintroduces the two changes that were reverted in PR2. Additionally:
- Since the cached XML for a fixture can contain multiple elements (e.g., both a schema and a fixture element), we add a comment indicating the number of elements so the restore includes the correct count for each fixture.
- This PR removes the step of aggregating all global fixtures into a single list with a comment indicating the total count. Instead, each component is returned as a separate element, which is appended to the restore in the same order.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
no FF.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Reproduced the bug locally and tested the change fixes the bug.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
A test was added that checks the fixture generator includes both the schema and fixture xml nodes. 

There exists tests that check: 
- fixture_generator creates the expected xml for non indexed fixtures
- fixture_generator correctly includes the fixtures that have been modified

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-5912]: https://dimagi.atlassian.net/browse/USH-5912?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

esoergel committed:
Add test demonstrating the issue

Jtang-1 committed:
bug: indexed data types contain more than one elements so we should not get just getting the first element.

Regardless, there was no reason to get the first element in the first place.

Jtang-1 committed:
Merge branch 'es/indexed-global-fixtures' of https://github.com/dimagi/commcare-hq into jt/lookup-fixture-bug-fix

Jtang-1 committed:
test: update so it's passing as expected

esoergel reviewed (approved):
The fix looks correct and makes sense to me.  The test failure appears related to an encoding issue, I think?  Like bytestrings vs strings

Jtang-1 committed:
Revert "Revert "Lookup Table Limit Restore to Recently Modified""

This reverts commit 7a2f67ad6d9fc1151a4e168b1fbeeb8b70cb7020.

Jtang-1 committed:
Merge branch 'jt/revert-revert-36091-jt/lookup-table-smart-sync' into jt/lookup-fixture-bug-fix

mkangia reviewed (commented):
Hey @Jtang-1 

Glad to see you got to the bottom of this.
I am confused by the last commit. Is that coming from a reviewed work?

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -182,17 +182,18 @@ def _get_or_cache_global_fixture(self, domain, global_type, overwrite_cache=Fals
                 if io_stream is None:
                     record_datadog_metric('generate', key)
-                    item = self._get_global_item(global_type, domain)
+                    items = self._get_global_items(global_type, domain)
```
No, I don't think there was an issue with the original code. Very originally, we built the xml for all the fixtures and cached it all together. With my PR, it was changed so we cached fixtures individually. 

The bug here was that in `_get_global_items `, I only returned the first index of `_get_fixtures`. If the lookup table has no indices set, `_get_fixtures` would just send back a list of 1 item. But with indices, it would include additional items for the `schema` element.

The diff shown in this PR is kinda messed up now, because the the problem PR was reverted.

mkangia replied:
I am a bit confused and getting lost in the context in the PR. But if you have an approval from @esoergel, then its cool.

Jtang-1 commented:
@mkangia I was trying to add back in the stuff that was reverted (the stuff in [c0dcedd](https://github.com/dimagi/commcare-hq/pull/36228/commits/c0dcedd76b0e115c684bd786d383e0cbf9fafc78)) I think I should have just cherrypicked it. But what i did was create a new branch that included that commit and then merged it into this branch.

Jtang-1 committed:
test: the xml byte string can contain multiple root elements (multiple fixture and schema elements) which is not valid XML.
XML must have one root element. This was resulting in a xml parsing error.

Jtang-1 committed:
Improve fixture restore handling for cached XML

Two main changes:

The cached XML for a fixture can contain multiple elements (e.g., both a schema and a fixture element). We add a comment indicating the number of elements so the restore includes the correct count for each fixture.

Removes the step of aggregating all global fixtures into a single list with a comment indicating the total count. Instead, each component is returned as a separate element, which is appended to the restore in the same order.
